# AIS - S20 - Exploratory Data Analysis Assignment

## John Bhanini

### EDA with movielens dataset

#### Outline:

     Loading the Data:
     
         Ratings Data:
         
            - Showing the raw ratings
            - Checking for missing values
            - Checking for duplicates
            
         Movies Data:
         
            - Showing the raw movies data
            - Checking for missing values
            - Checking for duplicates
            
     Duplicates are Hiding:
     
        - Duplicates in Movies
        - Duplicates in Ratings
        
     Outliers

In [109]:
from pathlib import Path
import pandas as pd

In [110]:
DATA_DIR = Path('../data/movielens/ml-latest')

RATINGS_FILEPATH = DATA_DIR / 'ratings.csv'
MOVIES_FILPATH = DATA_DIR / 'movies.csv'

# Loading the Data

## Ratings Data

### Showing the raw Ratings dataset.

In [111]:
ratings_raw = pd.read_csv(RATINGS_FILEPATH)
ratings_raw.sample(5)

userId  movieId  rating   timestamp
9548313    98475      110     4.0  1220499045
5502290    56675      708     3.0   842916234
4493120    46200     1127     3.0   862316191
10639851  109343      781     4.0  1015179906
5803401    59737     4361     3.0  1185670562

We then load the dataset in a new Ratings dataframe.

In [112]:
ratings = ratings_raw.copy()

In [113]:
ratings.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27753444 entries, 0 to 27753443
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 847.0 MB


### Checking for missing values

Checking for missing values in the data frame using a built in function. returns a boolean value for each column. If there is at least one missing value in that column, it returns True.

In [114]:
ratings.isnull().sum()

userId       0
movieId      0
rating       0
timestamp    0
dtype: int64

No missing values in Ratings data frame.

### Checking for duplicates

We check for duplicates our Ratings data frame using another built in function that returns the number of duplicates in the dataframe.

In [115]:
ratings.duplicated(keep=False).sum()

0

Here it returns 0.

## Movies Data

### Showing the raw Movies data set.

In [116]:
movies_raw = pd.read_csv(MOVIES_FILPATH)
movies_raw.sample(5)

movieId                                              title  \
52577   181183                When the Boys Meet the Girls (1965)   
13860    69255                             The Merry Widow (2007)   
30295   130247  That Gal... Who Was in That Thing: That Guy 2 ...   
25629   117370                               The Seven-Ups (1973)   
43825   162036                           The Bells Go Down (1943)   

                            genres  
52577           (no genres listed)  
13860                 Comedy|Drama  
30295                  Documentary  
25629  Action|Crime|Drama|Thriller  
43825                    Drama|War

We then load the dataset in a new Movies dataframe.

In [117]:
movies = movies_raw.copy()

In [118]:
movies.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 58098 entries, 0 to 58097
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   movieId  58098 non-null  int64 
 1   title    58098 non-null  object
 2   genres   58098 non-null  object
dtypes: int64(1), object(2)
memory usage: 1.3+ MB


### Checking for missing values

Checking for missing values in the data frame using a built in function. returns a boolean value for each column. If there is at least one missing value in that column, it returns True.

In [119]:
movies.isnull().sum()

movieId    0
title      0
genres     0
dtype: int64

No missing values in Movies data frame.

### Checking for duplicates

We check for duplicates in our Movies data frame using another built in function that returns the number of duplicates.

In [120]:
movies.duplicated().sum()

0

Here it returns 0.

# Duplicates are Hiding :)

## Duplicates in Movies 

We previously checked for duplicates both in our Movies & Ratings data frames. Although we got 0 both times, that doesn't quite mean that there are no duplicates. The function we used checks for duplicates according to the MovieID. Indeed, according to the MovieID, there are no duplicates in our data sets. But if you look closely at the next section, you will find that duplicates in fact exist.

In fact, the duplicates in our datasets exist not based on MovieID but based on the Movie Title.

In [121]:
movies.duplicated(subset=['title', 'genres']).sum()
duplicates = movies[movies.duplicated(subset=['title', 'genres'], keep=False)]
duplicates.title.values

array(['Dracula (1931)', 'Little Man (2006)', 'Offside (2006)',
       'Berlin Calling (2008)', 'Seven Years Bad Luck (1921)',
       'Home (2008)', 'Offside (2006)', 'Darling (2007)',
       'Girl, The (2012)', 'Girl, The (2012)', 'Beneath (2013)',
       'Clear History (2013)', 'Johnny Express (2014)', 'Beneath (2013)',
       'Clear History (2013)', 'Johnny Express (2014)', 'Darling (2007)',
       'Macbeth (2015)', 'Home (2008)', 'Macbeth (2015)',
       'Seven Years Bad Luck (1921)', 'Detour (2017)',
       'Little Man (2006)', 'Lucky (2017)', 'Dracula (1931)',
       'Lucky (2017)', 'Detour (2017)', 'Berlin Calling (2008)'],
      dtype=object)

We can see that 14 Movie Titles are duplicated.

In [122]:
duplicates

movieId                        title                   genres
2560      2644               Dracula (1931)                   Horror
11173    46865            Little Man (2006)                   Comedy
11412    48682               Offside (2006)             Comedy|Drama
13504    66511        Berlin Calling (2008)             Comedy|Drama
14038    70155  Seven Years Bad Luck (1921)                   Comedy
15637    79254                  Home (2008)                    Drama
15902    80330               Offside (2006)             Comedy|Drama
18867    93279               Darling (2007)                    Drama
19914    97773             Girl, The (2012)                    Drama
20835   101212             Girl, The (2012)                    Drama
21655   104035               Beneath (2013)                   Horror
21691   104155         Clear History (2013)                   Comedy
23834   111519        Johnny Express (2014)  Animation|Comedy|Sci-Fi
25046   115777               Beneath (2013)                   Horror
27572   122940         Clear History (2013)                   Comedy
29852   128991        Johnny Express (2014)  Animation|Comedy|Sci-Fi
30226   130062               Darling (2007)                    Drama
33001   136564               Macbeth (2015)                    Drama
36172   143978                  Home (2008)                    Drama
38804   150310               Macbeth (2015)                    Drama
44387   163246  Seven Years Bad Luck (1921)                   Comedy
46904   168774                Detour (2017)                 Thriller
48620   172427            Little Man (2006)                   Comedy
51339   178401                 Lucky (2017)                    Drama
52192   180205               Dracula (1931)                   Horror
52644   181329                 Lucky (2017)                    Drama
52795   181655                Detour (2017)                 Thriller
57269   191775        Berlin Calling (2008)             Comedy|Drama

This creates inconcistency in our data.

We can clearly see below that for the same Movie Title, more than one entry exists with different MovieID and the same Genre.

In [123]:
for movie_title in duplicates.title.unique():
    display(movies[movies.title == movie_title])

movieId           title  genres
2560      2644  Dracula (1931)  Horror
52192   180205  Dracula (1931)  Horror

movieId              title  genres
11173    46865  Little Man (2006)  Comedy
48620   172427  Little Man (2006)  Comedy

movieId           title        genres
11412    48682  Offside (2006)  Comedy|Drama
15902    80330  Offside (2006)  Comedy|Drama

movieId                  title        genres
13504    66511  Berlin Calling (2008)  Comedy|Drama
57269   191775  Berlin Calling (2008)  Comedy|Drama

movieId                        title  genres
14038    70155  Seven Years Bad Luck (1921)  Comedy
44387   163246  Seven Years Bad Luck (1921)  Comedy

movieId        title genres
15637    79254  Home (2008)  Drama
36172   143978  Home (2008)  Drama

movieId           title genres
18867    93279  Darling (2007)  Drama
30226   130062  Darling (2007)  Drama

movieId             title genres
19914    97773  Girl, The (2012)  Drama
20835   101212  Girl, The (2012)  Drama

movieId           title  genres
21655   104035  Beneath (2013)  Horror
25046   115777  Beneath (2013)  Horror

movieId                 title  genres
21691   104155  Clear History (2013)  Comedy
27572   122940  Clear History (2013)  Comedy

movieId                  title                   genres
23834   111519  Johnny Express (2014)  Animation|Comedy|Sci-Fi
29852   128991  Johnny Express (2014)  Animation|Comedy|Sci-Fi

movieId           title genres
33001   136564  Macbeth (2015)  Drama
38804   150310  Macbeth (2015)  Drama

movieId          title    genres
46904   168774  Detour (2017)  Thriller
52795   181655  Detour (2017)  Thriller

movieId         title genres
51339   178401  Lucky (2017)  Drama
52644   181329  Lucky (2017)  Drama

This will surely affect our Ratings too.

In [124]:
ratings[ratings.movieId == 46865]

userId  movieId  rating   timestamp
34191        332    46865     2.5  1160367907
124859      1215    46865     1.0  1185238454
227128      2231    46865     0.5  1191651526
413108      4208    46865     4.0  1221509439
475122      4836    46865     0.5  1342553315
...          ...      ...     ...         ...
27534644  280901    46865     4.0  1220666505
27555788  281096    46865     0.5  1166033444
27588229  281499    46865     1.5  1372546405
27661277  282242    46865     2.0  1325466212
27705218  282676    46865     3.0  1266952286

[455 rows x 4 columns]

To deal with this, we have to first choose one of the duplicated entries as the main one with it's Movie ID. We need to then replace the all the duplicates of a same movie with the one we chose both in the Movie dataframe and the Ratings data frame.
To do this we need to define some functions:
- replace_movie_id_in_ratings_dataframe: Replacing the MovieID in Ratings data frame with the one we chose as the main one.
- remove_old_movie_ids_from_movies_dataframe: Removing the duplicate values form our Movies data frame.
- handler_function_f: Handles both of our previous functions. Helps us preform the operation smoother. You can think of this as the orchestra conductor. 

In [125]:
def replace_movie_id_in_ratings_dataframe(dataframe: pd.DataFrame, new_movie_id: int, old_movie_id: int) \
    -> pd.DataFrame:
    dataframe.loc[dataframe.movieId == old_movie_id, 'movieId'] = new_movie_id
    return dataframe

In [126]:
def remove_old_movie_ids_from_movies_dataframe(dataframe: pd.DataFrame, old_movie_ids: [int]) -> pd.DataFrame:
    target_index = dataframe[dataframe.movieId.isin(old_movie_ids)].index
    dataframe = dataframe.drop(target_index)
    return dataframe

In [127]:
def handler_function_f(movies_dataframe: pd.DataFrame, ratings_dataframe: pd.DataFrame) -> (pd.DataFrame, pd.DataFrame):
    duplicates = movies_dataframe[movies_dataframe.duplicated(subset=['title', 'genres'], keep=False)]
    old_movie_ids = []
    for movie_title in duplicates.title.unique():
        movie_ids = duplicates[duplicates.title == movie_title].movieId.values
        new_movie_id, old_movie_id = movie_ids
        ratings_dataframe = replace_movie_id_in_ratings_dataframe(ratings_dataframe, new_movie_id, old_movie_id)
        old_movie_ids.append(old_movie_id)
    movies_dataframe = remove_old_movie_ids_from_movies_dataframe(movies_dataframe, old_movie_ids)
    return movies_dataframe, ratings_dataframe

Let's now test our functions:

Replacing a MovieID in our Ratings data frame:

In [128]:
ratings = replace_movie_id_in_ratings_dataframe(ratings, 46865, 172427)
(ratings.movieId == 172427).sum()

0

Works Well.

Removing an old MovieID from our Movies dataframe:

In [129]:
remove_old_movie_ids_from_movies_dataframe(movies, [1, 2, 3, 4])

movieId                                    title  \
4            5       Father of the Bride Part II (1995)   
5            6                              Heat (1995)   
6            7                           Sabrina (1995)   
7            8                      Tom and Huck (1995)   
8            9                      Sudden Death (1995)   
...        ...                                      ...   
58093   193876                  The Great Glinka (1946)   
58094   193878  Les tribulations d'une caissière (2011)   
58095   193880                 Her Name Was Mumu (2016)   
58096   193882                             Flora (2017)   
58097   193886                              Leal (2018)   

                              genres  
4                             Comedy  
5              Action|Crime|Thriller  
6                     Comedy|Romance  
7                 Adventure|Children  
8                             Action  
...                              ...  
58093             (no genres listed)  
58094                         Comedy  
58095                          Drama  
58096  Adventure|Drama|Horror|Sci-Fi  
58097             Action|Crime|Drama  

[58094 rows x 3 columns]

Also works well.

And if we run our handler_function_f:

In [130]:
movies_df, ratings_df = handler_function_f(movies, ratings)

We can see by substracting the number of entries in our Movies data frame from the number of entries in the original movies_raw dataset that we have succesfully reduced 14 entries (the duplicated values).

In [131]:
len(movies_raw) - len(movies_df)

14

Similarly, by substracting the number of entries in our Ratings data frame from the original number of entries in ratings_raw dataset, we see that the ratings are unchanged.

This is because we don't remove any entries from our Ratings data frame, instead we replace them as discussed before.

In [132]:
len(ratings_raw) - len(ratings_df)

0

## Duplicates in Ratings

Let us now take a closer look at the updated ratings dataframe we generated.

In [133]:
ratings_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27753444 entries, 0 to 27753443
Data columns (total 4 columns):
 #   Column     Dtype  
---  ------     -----  
 0   userId     int64  
 1   movieId    int64  
 2   rating     float64
 3   timestamp  int64  
dtypes: float64(1), int64(3)
memory usage: 847.0 MB


In [134]:
ratings_df

userId  movieId  rating   timestamp
0              1      307     3.5  1256677221
1              1      481     3.5  1256677456
2              1     1091     1.5  1256677471
3              1     1257     4.5  1256677460
4              1     1449     4.5  1256677264
...          ...      ...     ...         ...
27753439  283228     8542     4.5  1379882795
27753440  283228     8712     4.5  1379882751
27753441  283228    34405     4.5  1379882889
27753442  283228    44761     4.5  1354159524
27753443  283228    54286     4.5  1354159718

[27753444 rows x 4 columns]

In [135]:
ratings_df[ratings_df.duplicated(['userId','movieId'],keep=False)]

userId  movieId  rating   timestamp
6213836    63786   104155     0.5  1490157153
6214019    63786   104155     0.5  1490157134
6575743    67522     2644     3.0  1522388123
6576057    67522     2644     4.0  1522436661
7128279    73145   104155     4.0  1384276985
7128522    73145   104155     4.0  1421188224
8204118    84432   104155     3.5  1390947133
8204136    84432   104155     3.0  1432918104
10394207  106859    70155     3.0  1500908547
10394847  106859    70155     3.0  1500908560
10397896  106873   104155     4.0  1376975238
10398018  106873   104155     4.5  1436758238
10862664  111688   104155     4.5  1402494730
10862692  111688   104155     4.5  1477450047
12010931  123100    97773     3.0  1450166863
12011347  123100    97773     3.0  1450164192
12243097  125365   111519     4.5  1436452325
12243194  125365   111519     4.5  1427199640
15158896  154880   104155     4.0  1398187007
15158951  154880   104155     3.5  1512964435
15264429  155976    48682     3.5  1267648381
15264776  155976    48682     3.5  1463268492
15653605  159846    66511     2.0  1499318921
15654247  159846    66511     2.0  1534046590
16408991  167835   104155     3.0  1388380607
16409115  167835   104155     3.0  1457545603
17920236  183200   104035     0.5  1475292730
17920390  183200   104035     1.0  1475285150
19174782  195628   104155     3.0  1381140059
19174990  195628   104155     3.0  1530661998
19242274  196323     2644     5.0  1451170687
19243096  196323     2644     5.0  1530938260
21701689  221697   178401     4.0  1527358587
21701699  221697   178401     2.0  1527358755
22505207  230183    97773     3.5  1501655001
22505255  230183    97773     3.5  1464937962
25465019  260133   104155     3.5  1378778895
25465161  260133   104155     2.5  1505977593
26353900  269098   104155     3.5  1377396957
26354146  269098   104155     3.5  1496019855
27242840  278022    46865     3.5  1494713647
27244405  278022    46865     4.0  1494717970

In [136]:
ratings_df.duplicated(['userId','movieId']).sum()

21

We notice that we have some duplicates. 21 to be exact. Let us further examine this.

In [137]:
ratings_df[ratings_df.duplicated(['userId','movieId','rating'],keep=False)]

userId  movieId  rating   timestamp
6213836    63786   104155     0.5  1490157153
6214019    63786   104155     0.5  1490157134
7128279    73145   104155     4.0  1384276985
7128522    73145   104155     4.0  1421188224
10394207  106859    70155     3.0  1500908547
10394847  106859    70155     3.0  1500908560
10862664  111688   104155     4.5  1402494730
10862692  111688   104155     4.5  1477450047
12010931  123100    97773     3.0  1450166863
12011347  123100    97773     3.0  1450164192
12243097  125365   111519     4.5  1436452325
12243194  125365   111519     4.5  1427199640
15264429  155976    48682     3.5  1267648381
15264776  155976    48682     3.5  1463268492
15653605  159846    66511     2.0  1499318921
15654247  159846    66511     2.0  1534046590
16408991  167835   104155     3.0  1388380607
16409115  167835   104155     3.0  1457545603
19174782  195628   104155     3.0  1381140059
19174990  195628   104155     3.0  1530661998
19242274  196323     2644     5.0  1451170687
19243096  196323     2644     5.0  1530938260
22505207  230183    97773     3.5  1501655001
22505255  230183    97773     3.5  1464937962
26353900  269098   104155     3.5  1377396957
26354146  269098   104155     3.5  1496019855

In [138]:
ratings_df.duplicated(['userId','movieId','rating']).sum()
length1 = len(ratings_df)

We can clearly see that some users have rated the same movie with the same rating more than once. We have 13 instances of this. This will cause inconsitency later on. We need to remove them. No need for complicated functions at this time, the pandas built-in df.drop will do the job.

In [139]:
ratings_1 = ratings_df.drop_duplicates(subset=['userId','movieId','rating'])
ratings_1[ratings_1.duplicated(['userId','movieId'],keep=False)]

userId  movieId  rating   timestamp
6575743    67522     2644     3.0  1522388123
6576057    67522     2644     4.0  1522436661
8204118    84432   104155     3.5  1390947133
8204136    84432   104155     3.0  1432918104
10397896  106873   104155     4.0  1376975238
10398018  106873   104155     4.5  1436758238
15158896  154880   104155     4.0  1398187007
15158951  154880   104155     3.5  1512964435
17920236  183200   104035     0.5  1475292730
17920390  183200   104035     1.0  1475285150
21701689  221697   178401     4.0  1527358587
21701699  221697   178401     2.0  1527358755
25465019  260133   104155     3.5  1378778895
25465161  260133   104155     2.5  1505977593
27242840  278022    46865     3.5  1494713647
27244405  278022    46865     4.0  1494717970

Preform a quick check to see if our operation works:

In [140]:
len(ratings_df)-len(ratings_1)

13

It does!

Originally, we had 21 duplicates in our data frame, after this step where we removed 13 we should only be left with 8.

In [141]:
ratings_1.duplicated(['userId','movieId']).sum()

8

Let us check them out:

In [142]:
ratings_1[ratings_1.duplicated(['userId','movieId'],keep=False)]

userId  movieId  rating   timestamp
6575743    67522     2644     3.0  1522388123
6576057    67522     2644     4.0  1522436661
8204118    84432   104155     3.5  1390947133
8204136    84432   104155     3.0  1432918104
10397896  106873   104155     4.0  1376975238
10398018  106873   104155     4.5  1436758238
15158896  154880   104155     4.0  1398187007
15158951  154880   104155     3.5  1512964435
17920236  183200   104035     0.5  1475292730
17920390  183200   104035     1.0  1475285150
21701689  221697   178401     4.0  1527358587
21701699  221697   178401     2.0  1527358755
25465019  260133   104155     3.5  1378778895
25465161  260133   104155     2.5  1505977593
27242840  278022    46865     3.5  1494713647
27244405  278022    46865     4.0  1494717970

We can see that the 8 instances are times where the same user has rated the same movie with different values. A way to deal with this would be to take the average user rating value for the same movie.But i don't think that this is necessary because this will happen both ways later on.

## Outliers

Let's look at our Ratings from a statistical point of view now.

In [143]:
ratings_1.describe()

userId       movieId        rating     timestamp
count  2.775343e+07  2.775343e+07  2.775343e+07  2.775343e+07
mean   1.419420e+05  1.848780e+04  3.530445e+00  1.193122e+09
std    8.170741e+04  3.510216e+04  1.066353e+00  2.160482e+08
min    1.000000e+00  1.000000e+00  5.000000e-01  7.896520e+08
25%    7.117600e+04  1.097000e+03  3.000000e+00  9.986048e+08
50%    1.420220e+05  2.716000e+03  3.500000e+00  1.174256e+09
75%    2.124590e+05  7.150000e+03  4.000000e+00  1.422744e+09
max    2.832280e+05  1.938860e+05  5.000000e+00  1.537945e+09

Using the .describe we can get some info regarding our data frame but not very usefull. We have to do some more manipulations ourself.

Let's try to see the number  of ratings each movie has:

In [144]:
ratings_1.groupby('movieId').count()

userId  rating  timestamp
movieId                           
1         68469   68469      68469
2         27143   27143      27143
3         15585   15585      15585
4          2989    2989       2989
5         15474   15474      15474
...         ...     ...        ...
193876        1       1          1
193878        1       1          1
193880        1       1          1
193882        1       1          1
193886        2       2          2

[53877 rows x 3 columns]

It's clear that our distribution has many outliers. For example: the first movie we see  recieved 68469 ratingss. Another movie recieved 15474 ratings. With some movies recieving 1 or 2 ratings only. I'm not sure how to handle this exactly, but i think grouping movies by the number of ratings recieved might be a good approach. Not the exact number but a close interval of ratings. 